### DISCUSSION TITLE
Multiple Placeholders

### DISCUSSION THREAD

**Discussion by swnydick** [davidgohel/officer#415] [Q&A] [answered]
I have a question. Why did you remove the ability to have multiple placeholders on a powerpoint slide (or even to get multiple locations for those placeholders)? I actually use that quite a lot when I have text that I know to be the exactly the same in different places on the slide (which is very common given what I do and what I automate). It helps me keep my code clean and allows me to easily debug/check the template and code to make sure everything is going in the right place.

I think this used to be allowed in a very old version of the package, but a while ago, you removed the ability to do it in ph_with for placeholders. However, I was able to get around that limitation by using fortify location directly and then using the locations to place the text in the placeholders. However, in a recent version of officer, you made fortify location (specifically `fortify_location.location_label`) spit out an error if there are multiple placeholders. Note that the code is the same and it works if I somehow can remove that `if(nrow(props) > 1) stop()`. My only options are:

1. Update all of my templates to have placeholders be different even if the text should be the same and then somehow find a way of using this update in the code, which seems error prone.
2. Essentially monkey patch the old version of this function into the package.
3. Use an old version of officer. This is what I am currently doing and only ran into the problem when that upgraded on a newer computer.

This is oddly an incredibly important feature for how I program automated reports, even if it is a bit off-label.

**Comment by davidgohel** [comment]
I don't understand what you are referring to, there is no limitation on the ability to have multiple placeholders on a PowerPoint slide. 
The pptx API is stable since 2 or 3 years and I don't remember having changed something like that. 

Could you share a some code that help me to figure out how to reproduce?

**Comment by swnydick** [comment]
Sorry. Not having multiple placeholders in general, having multiple placeholders with the same name on the same slide. Like if I want to put "Copyright 2022, Me" multiple places on a pptx slide, I will have the placeholder name "copyright" in every place where I want to put that text.

**Comment by davidgohel** [comment]
The label must be a unique id since the function `ph_location_label()` exists, at least it exists since Nov 2018, see https://github.com/davidgohel/officer/commit/5d11c398092b56a3a28e7afe1bad59c0441085a2#diff-1bc0a6dddbc3e31db1ddb5d62416a0fa83d6baae188cc84d143ce7a0ae54e0b4R124

Because of this issue [363](https://github.com/davidgohel/officer/issues/363), I changed the mechanism, from `warning()` to `stop()` a year ago: 
https://github.com/davidgohel/officer/commit/04bf439409d6c5837671d8447bdbbddfe13fec59


The warning was there since the beginning. How could you have use that function with multiple similar id/labels? Please share a code, it makes no sense to me.

**Comment by swnydick** [comment]
You're right. I've always had to cycle through it. I just checked my old code. This was a long time ago, and I was using your old interface with `layout_properties` and `ph_with_text` to update labels. When you finalized the API and from Version 0.3.14, I had to change my code to:

```
add_object_by_ph_ <- function(pptx,
                              object,
                              placeholder,
                              use_first_loc = FALSE,
                              ...){
  
  # if we are only doing the first location, use standard function
  if(isTRUE(use_first_loc)){
    pptx <- ph_with(
      pptx,
      value    = object,
      location = ph_location_label(ph_label = placeholder),
      ...
    )
    
    return(pptx)
  }
  
  # pull out all locations with label and create list of them in order
  ph_loc   <- suppressWarnings({
    ph_location_label(ph_label = placeholder,
                      x        = pptx) %>%
    fortify_location(doc       = pptx)
  })

  # make sure that the list is inverted so that each element is row
  ph_loc   <- do.call(what = Map,
                      args = c(list(f = list), ph_loc))
    
  # run ph_with across all of the instances of the label
  for(loc in ph_loc){
    
    # due to changes in new version of officer ph_location
    loc         <- do.call(args = loc,
                           what = ph_location_identity)
    class(loc)  <- "location_manual"
      
    pptx <- ph_with(
      pptx,
      value    = object,
      location = loc,
      ...
    )
  }

  return(pptx)  
}
```

That said, it used to give a warning and now gives an error in the `ph_location_label` function. I agree this is an off-label use, but I wasn't sure how else to do this, as multiple placeholders with the same name on a slide keeps my powerpoint and code clean. I guess I was wondering why you turned it from a warning to an error. I see that it's because you can accidentally have the same placeholder on the slide (rather than intentionally, as I do). I agree it would probably be more confusing than it's worth to have the option of finding the location for multiple placeholders given the typical use case.

**Reply by markheckmann** [reply]
Would a function `ph_location_id` come in handy for your use case? Then, you could get the ids of all placeholders with a given label and cycle through them as you did before, but referencing their location via id. **Would that do the job?**

**Reply by swnydick** [reply]
I think that sounds like a reasonable solution!

**Reply by markheckmann** [reply]
Great. See issue #599. I will implement it over the next weeks.

**Reply by markheckmann** [reply, marked as answer]
@swnydick, as @davidgohel points out [here](https://github.com/davidgohel/officer/issues/599#issuecomment-2327466821), `ph_location_type()` already takes the `id` as a parameter to reference a ph.

 Can this thread be marked as answered or is there more functionality needed for your use case?
